core/graphelier-service/models/orderbooks.go

```diff
@@ -267,3 +267,31 @@ func (orderbook *Orderbook) ApplyMessagesToDeltabook(messages []*Message, numMes
 
 	return
 }
+
+// TopBookPerXNano : Returns a list of the top book at every x nanosecond after applying each individual message in a list
+func (orderbook *Orderbook) TopBookPerXNano(messages []*Message, x uint64) (topbook []*Point) {
+	topbookLengthIndex := 0
+	initialTimestamp := messages[0].Timestamp
+	for _, message := range messages {
+		singleMsgSlice := []*Message{message}
+		orderbook.ApplyMessagesToOrderbook(singleMsgSlice)
+		if (message.Timestamp-initialTimestamp)%x == 0 {
+			tb := &Point{}
+			tb.Timestamp = message.Timestamp
+			if len(orderbook.Bids) != 0 {
+				tb.BestBid = orderbook.Bids[0].Price
+			}
+			if len(orderbook.Asks) != 0 {
+				tb.BestAsk = orderbook.Asks[0].Price
+			}
+			if topbookLengthIndex > 0 && message.Timestamp == topbook[topbookLengthIndex-1].Timestamp {
+				topbook[topbookLengthIndex-1] = tb
+			} else {
+				topbook = append(topbook, tb)
+				topbookLengthIndex++
+			}
```

Comment:
What is this for?

Reply:
it is my attempt to replace the best bid/ask if we have multiple messages at the same timestamp